unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/NFTs/BuilderInWorldNFTController.cs

```diff
@@ -0,0 +1,146 @@
+using DCL.Configuration;
+using DCL.Helpers.NFT;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Text;
+using UnityEngine;
+
+public class BuilderInWorldNFTController 
+{
+    public event System.Action OnNFTUsageChange;
+    NFTOwner nFTOwner;
+
+    Coroutine fechNFTsCor;
+
+    static BuilderInWorldNFTController instance;
+
+    List<NFTInfo> nFTsAlreadyInUse = new List<NFTInfo>();
+
+    bool desactivateNFT = false;
+
+    public static BuilderInWorldNFTController i
+    {
+        get
+        {
+            if (instance == null)
+            {
+                instance = new BuilderInWorldNFTController();
+            }
+
+            return instance;
+        }
+    }
+ 
+    public void Start()
+    {
+        UserProfile userProfile = UserProfile.GetOwnUserProfile();
+        userProfile.OnUpdate += (x) => GetNFTsFromOwner();
+    }
+
+    public void ClearNFTs()
+    {
+        nFTsAlreadyInUse.Clear();
+    }
+
+    public bool IsNFTInUse(string id)
+    {
+        if (desactivateNFT)
+            return false;
+
+        foreach(NFTInfo info in nFTsAlreadyInUse)
+        {
+            if (info.assetContract.address == id)
+                return true;
+        }
+        return false;
+    }
+
+    public void StopUsingNFT(string id)
+    {
+        if (desactivateNFT)
+            return;
+
+        foreach (NFTInfo info in nFTOwner.assets)
+        {
+            if (info.assetContract.address != id) continue;
+            if (!nFTsAlreadyInUse.Contains(info)) continue;
+
+            nFTsAlreadyInUse.Remove(info);
+            OnNFTUsageChange?.Invoke();
+        }
+    }
+
+    public void UseNFT(string id)
+    {
+        if (desactivateNFT)
+            return;
+
+        foreach (NFTInfo info in nFTOwner.assets)
+        {
+            if (info.assetContract.address != id) continue;
+            if (nFTsAlreadyInUse.Contains(info)) continue;
+
+            nFTsAlreadyInUse.Add(info);
+            OnNFTUsageChange?.Invoke();
+
+        }
+    }
+   
+    public List<SceneObject> GetNFTsAsSceneObjects()
+    {
+        List<SceneObject> sceneObjects = new List<SceneObject>();
+
+        if (desactivateNFT ||
+            nFTOwner.assets == null)
+            return sceneObjects;
+
+        foreach(NFTInfo nFTInfo in nFTOwner.assets)
+        {
+            sceneObjects.Add(NFTInfoToSceneObject(nFTInfo));
+        }
+        return sceneObjects;
+    }
+
+    void GetNFTsFromOwner()
```

Comment:
What's this class lifecycle? I see many confusing things here:

- We have a Start() method
- We have this `GetNFTsFromOwner()` that restarts a coroutine each time its called.

To clarify this I suggest:
- Use constructor instead of `Start`, or at the very least rename `Start()` to `Initialize()`.
- Rename `GetNFTsFromOwner` to `FetchNFTsFromOwner` to ensure the method intent is more clear.
- Make 100% sure that interrupting the former coroutine can't leave the application in a invalid state, if we find the interruption is not safe, we can just implement an ongoing coroutine that enqueues the NFT wanted updates or something similar.

Reply:
- I have renamed it to Initialize since it should start fetching NFT at the activation of the Buidler In World

- Renamed

- I don't quite understand this, why would interrupt this coroutine level the application in an invalid state?